piccolo/tablestorage.py

```diff
@@ -0,0 +1,111 @@
+import asyncio
+import typing as t
+
+from piccolo.apps.schema.commands.generate import get_output_schema
+from piccolo.table import Table, TableMetaclass
+
+
+class Immutable(object):
+    def _immutable(self, *arg, **kw) -> TypeError:
+        raise TypeError("%s object is immutable" % self.__class__.__name__)
+
+    __delitem__ = __setitem__ = __setattr__ = _immutable
+
+
+class ImmutableDict(Immutable, dict):
+    """A dictionary that is not publicly mutable."""
+
+    clear = pop = popitem = setdefault = update = Immutable._immutable
+
+    def __new__(cls, *args):
+        new = dict.__new__(cls)
+        return new
+
+    def copy(self):
+        raise NotImplementedError(
+            "an immutabledict shouldn't need to be copied.  use dict(d) "
+            "if you need a mutable dictionary."
+        )
+
+    def __reduce__(self):
+        return ImmutableDict, (dict(self),)
+
+    def _insert_item(self, key, value) -> None:
+        """
+        insert an item into the dictionary directly.
+        """
+        dict.__setitem__(self, key, value)
+
+    def _delete_item(self, key) -> None:
+        """
+        Delete an item from dictionary directly.
+        """
+        dict.__delitem__(self, key)
+
+    def __repr__(self):
+        return f"ImmutableDict({dict.__repr__(self)})"
+
+
+class Singleton(type):
+    """
+    A metaclass that creates a Singleton base class when called.
+    """
+
+    _instances: t.Dict = {}
+
+    def __call__(cls, *args, **kwargs):
+        if cls not in cls._instances:
+            cls._instances[cls] = super(Singleton, cls).__call__(
+                *args, **kwargs
+            )
+        return cls._instances[cls]
+
+
+class TableStorage(metaclass=Singleton):
+    """
+    A singleton object to store reflected tables
+    """
+
+    def __init__(self):
+        self.tables = ImmutableDict()
+
+    async def reflect(
+        self,
+        schema_name: str = "public",
+        if_not_exists: t.Optional[bool] = False,
+    ) -> None:
+        output_schema = await get_output_schema(schema_name=schema_name)
+        add_tables = [
+            self._add_table(
+                schema_name=schema_name,
+                table=table,
+                if_not_exists=if_not_exists,
+            )
+            for table in output_schema.tables
+        ]
+        await asyncio.gather(*add_tables)
+
+    async def _add_table(
+        self,
+        schema_name: str,
+        table: t.Type[Table],
+        if_not_exists: t.Optional[bool] = False,
```

Comment:
Do you think we need `if_not_exists` here? If so, perhaps the type should just be `bool`?

Reply:
I agree. This is not useful in here. This should be handled way before this point to have an at least a performance advantage. Some kind of caching mechanism could solve the problem too. 
The point is to decrease the overhead of creating all the objects from beginning when an update to the `TableStorage` is needed.
If explanation is unclear, please tell me and I'll try to explain in another way.

Reply:
I see what you're saying. Maybe it's a useful feature - it depends if people will reflect the same database multiple times.